server/src/pull/pull.ts

```diff
@@ -0,0 +1,252 @@
+import {z} from 'zod';
+import type {PatchOperation, PullResponse, PullResponseOKV1} from 'replicache';
+import {transact} from '../pg';
+import {getClientGroupForUpdate, putClientGroup, searchClients} from '../data';
+import type Express from 'express';
+import {hasNextPage, isPageEmpty, readNextPage} from './next-page';
+import {
+  CVR,
+  dropCVREntries,
+  getCVR,
+  putCVR,
+  recordDeletes,
+  recordUpdates,
+} from './cvr';
+import {PartialSyncState, PARTIAL_SYNC_STATE_KEY} from 'shared';
+
+const cookieSchema = z.object({
+  order: z.number(),
+  clientGroupID: z.string(),
+});
+
+export type Cookie = z.infer<typeof cookieSchema>;
+
+export const pullRequest = z.object({
+  clientGroupID: z.string(),
+  cookie: z.union([cookieSchema, z.null()]),
+});
+
+export async function pull(
+  requestBody: Express.Request,
+): Promise<PullResponse> {
+  const start = performance.now();
+  console.log(`Processing pull`, JSON.stringify(requestBody, null, ''));
+  const pull = pullRequest.parse(requestBody);
+  const {clientGroupID} = pull;
+
+  const {clientChanges, nextPage, prevCVR, nextCVR} = await transact(
+    async executor => {
+      // Get a write lock on the client group first to serialize with other
+      // requests from the CG and avoid deadlocks.
+      const baseClientGroupRecord = await getClientGroupForUpdate(
+        executor,
+        clientGroupID,
+      );
+      const prevCVR = pull.cookie
+        ? await getCVR(executor, pull.cookie.clientGroupID, pull.cookie.order)
+        : undefined;
+      const baseCVR: CVR = prevCVR ?? {
+        clientGroupID,
+        clientVersion: 0,
+        order: 0,
+      };
+
+      // Drop any cvr entries greater than the order we received.
```

Comment:
> What happens if we drop cvr enteries for a cvrversion (i.e. order) and then later getting a pull with that cvrversion (i.e. order)?

We'd compute a new patch from head. The new patch would be recorded with the next cvrversion.

Does that work with respect to Replicache's expectations?

Reply:
Elsewhere @grgbkr suggested that this implies that when we the pull endpoint receives an old cookie, it should _not_ delete newer entries.

Instead, it should do the diff against client_view_entries that existed as of the received cookie and effectively fast-forward itself to the current state of the database.

This implies that this pull chunk might be bigger than desired but I think that's a good tradeoff.

Reply:
> Instead, it should do the diff against client_view_entries that existed as of the received cookie and effectively fast-forward itself to the current state of the database.

Which we can do while deleting new entries. On Friday, I thought we landed on:

1. keep the delete
2. remove the <= from `findUnsentItems`
3. " diff against client_view_entries that existed as of the received cookie" which, to me, means: "re-run findUnsentItems with the received cookie"

> Since we dropped the cvr entries we don't know what exactly the client has in its client view to compute a diff against, right?

We still have the entries as of the received cookie. I may be misunderstanding but, to me, if the client sends a cookie then that is telling us what they have. They have all data sent up to and including that cookie value but nothing later. Of course the client _could_ have something later since there _could_ be two tabs fighting but the server can't differentiate between that case or packets being dropped and causing the client to send old cookies.

We don't have to delete new entries but I think that puts us back at needing to make the entries immutable.

Reply:
I think I see the issue you're getting at (https://github.com/rocicorp/repliear/pull/158#discussion_r1415766295). Let me think on this a bit more 🤔